### PR TITLE
Joshua Jordal

### PR DESCRIPTION
**Pull request by JoshuaJordal**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,66 +1,3 @@
...
-* If `README` is not clear, missing or not edited, or `Dockerfile` is not updated, up to 15 points will be docked off.
-
-* If none of the functionalities work, 10 points will be given assuming `credentials.ini` is submitted with the correct repo URL, and `Dockerfile` builds and runs without any errors. 
-
-## Authors
-
-Michal Young, Ram Durairajan. Updated by Ali Hassani.
+Joshua Jordal
+jjordal@uoregon.edu
+This application calculates the opening and closing time of checkpoints on a race
```
Comment (CalebSutherland) [inline README.md:3]: You should add a description of how the algorithms are calculated and directions on how to start the app

File: brevets/acp_times.py [hunk 2]
```diff
@@ -1,44 +1,103 @@
...
+        return [15, 34]
+    elif(dist <= 400):
+        return [15, 32]
+    elif(dist <= 600):
+        return[15, 30]
+    elif(dist <= 1000):
+        return [11.428, 28]
+    else:
+        return [13.333, 26]
+
```
Comment (CalebSutherland) [inline brevets/acp_times.py:26]: If you wanted to add more distances, you would have to keep adding elif satements, instead you could use a dictionary. Since you have the distance and differences list in both functions, you could add this data to it also. For example,
```
    MaxSpeed = {200: (200, 34), 300: (100, 32), 400: (100, 32), etc }
```

File: brevets/acp_times.py [hunk 3]
```diff
@@ -1,44 +1,103 @@
...
+           or 1000 (the only official ACP brevet distances)
+       brevet_start_time:  An arrow object
+    Returns:
+       An arrow object indicating the control open time.
+       This will be in the same time zone as the brevet start time.
+    """
+    distances = [200, 400, 600, 1000, 1300]
+    difference = [200, 200, 200, 400, 300]
+    time = 0
+    i = 0
```
Comment (CalebSutherland) [inline brevets/acp_times.py:43]: As said above, if you used a dictionary, seed and difference could be assigned outside the if statements

File: brevets/acp_times.py [hunk 4]
```diff
@@ -1,44 +1,103 @@
...
+    for dist in distances:
+      if control_dist_km <= 0:
+         break
+      if control_dist_km - dist > 0:
+         speeds = speed(dist)[1]
+         time += difference[i] / speeds
+      elif control_dist_km - dist == 0:
+         speeds = speed(dist)[1]
+         time += difference[i] / speeds
```
Comment (CalebSutherland) [inline brevets/acp_times.py:52]: first two statements do the same thing, you could change it to >= 0 to cover both cases

File: brevets/acp_times.py [hunk 5]
```diff
@@ -1,44 +1,103 @@
...
+         remainder = control_dist_km -  (0 if i == 0 else distances[i - 1])
+         speeds = speed(dist)[1]
+         time += remainder / speeds
+         break
+      i += 1
+    hour = trunc(time)
+    time = time - hour
+    minute = round(time * 60)
+    open_time = brevet_start_time.shift(hours=+hour, minutes=+minute)
+    return open_time
```
Comment (CalebSutherland) [inline brevets/acp_times.py:64]: Overall, the logic behind the open time algorithm looks good

File: brevets/acp_times.py [hunk 6]
```diff
@@ -1,44 +1,103 @@
...
+          (the only official ACP brevet distances)
+       brevet_start_time:  An arrow object
+    Returns:
+       An arrow object indicating the control close time.
+       This will be in the same time zone as the brevet start time.
+    """
+    distances = [200, 400, 600, 1000, 1300]
+    difference = [200, 200, 200, 400, 300]
+    time = 0
+    i = 0
```
Comment (CalebSutherland) [inline brevets/acp_times.py:81]: There should be a case for when the control distance is less than 60 that uses a separate algorithm that is described in the oddities section of the rules of the ACP brevet time calculator

File: brevets/acp_times.py [hunk 7]
```diff
@@ -1,44 +1,103 @@
...
+         remainder = control_dist_km -  (0 if i == 0 else distances[i - 1])
+         speeds = speed(dist)[0]
+         time += remainder / speeds
+         break
+      i += 1
+    hour = trunc(time)
+    time = time - hour
+    minute = round(time * 60)
+    close_time = brevet_start_time.shift(hours=+hour, minutes=+minute)
+    return close_time
```
Comment (CalebSutherland) [inline brevets/acp_times.py:102]: The close time algorithm looks good, but it should if the control is = to the brevet distance, the close time should be the official close time for the brevet which can be found on wikipedia. For example, 200km calculates to 13H20M but the official closing time is 13H30M

File: brevets/flask_brevets.py [hunk 8]
```diff
@@ -1,74 +1,76 @@
...
+    app.logger.debug("request.args: {}".format(request.args))
+    # FIXME!
+    # Right now, only the current time is passed as the start time
+    # and control distance is fixed to 200
+    # You should get these from the webpage!
+    open_time = acp_times.open_time(km, brevet, arrow.get(time)).format('YYYY-MM-DDTHH:mm')
+    close_time = acp_times.close_time(km, brevet, arrow.get(time)).format('YYYY-MM-DDTHH:mm')
+    result = {"open": open_time, "close": close_time}
+    return flask.jsonify(result=result)
+
```
Comment (CalebSutherland) [inline brevets/flask_brevets.py:66]: Looks good to me, only difference I had was setting time to str but I don't think it matters in this case

File: brevets/templates/calc.html [hunk 9]
```diff
@@ -1,176 +1,178 @@
...
+            // Right now, only the kilometers are passed
+            // you need to pass much more!
+            // View flask_brevets.py for details.
+            function calc_times(control) {
+              var km = control.find("input[name='km']").val();
+              var open_time_field = control.find("input[name='open']");
+              var close_time_field = control.find("input[name='close']");
+              var brevet = $('#brevet_dist_km').find(":selected").val();
+              var time = $('#begin_date').val()
+
```
Comment (CalebSutherland) [inline brevets/templates/calc.html:128]: I did the same thing here to find brevet dist and begin time

File: brevets/tests/test_acp_times.py [hunk 10]
```diff
@@ -1,11 +1,32 @@
...
+#from src.acp_times import open_time
+import arrow
+import collections
+collections.Callable = collections.abc.Callable
+from acp_times import *
+logging.basicConfig(format='%(levelname)s:%(message)s',
+                    level=logging.WARNING)
+log = logging.getLogger(__name__)
+
+
```
Comment (CalebSutherland) [inline brevets/tests/test_acp_times.py:18]: The only difference I would make would be to use a dictionary to test multiple checkpoints for one brevet. For example,
```
    start_time = arrow.get("2023-01-10 00:00", "YYYY-MM-DD HH:mm")
    dist = 200
    checkpoints = {
            0: (start_time, start_time.shift(hours=1)),
            20: (start_time.shift(minutes=35), start_time.shift(hours=2)),
            60: (start_time.shift(hours=1, minutes=46), start_time.shift(hours=4)),
            etc
        }
```
and then loop through it
```
for km, times in checkpoints.items():
    check_open, check_close = times
    assert(open_time(km, dist, start_time)) == check_open
    assert(close_time(km, dist, start_time)) == check_close
```

File: brevets/acp_times.py (hunk 3 above)
Comment (CalebSutherland) [inline brevets/acp_times.py:43]:
```suggestion
    i = 0
    if control_dist_km > brevet_dist_km:
        control_dist_km = brevet_dist_km
```

**Review by CalebSutherland — CHANGES_REQUESTED** [review]
The webpage works when I run it, and most of the calculations are spot on. I like how you coded your logic for the open and close algorithms.

There are only a few edge cases that were overlooked. If the control distance is > brevet distance, it should be set to the brevet distance, I added a suggestion and it should cover it. For the close time algorithm, it should use a separate algorithm to calculate the time for distance less than 60. It should also use the official ending times for controls set to the brevet distance. Adding a dictionary as I described to hold distances and speeds would also help if you wanted easily be able to add more distances in the future. The same thing goes for the tests if you wanted to easily add more cases.

Other than those edge cases and a few spots for cleaner code, it looks good!
